# SQL Skills Practice: Exploring the Chinook Database

## Introduction
In this guided project, I will be honing my SQL skills by exploring the Chinook database. The Chinook database, provided as a SQLite file called `chinook.db`, offers valuable insights into a digital media store's operations.

## Database Overview
The Chinook database schema encompasses tables representing various aspects of the digital media store, including customers, invoices, tracks, albums, and genres.

## Project Objectives
My goal is to leverage SQL queries to address specific business questions and extract actionable insights from the Chinook database. By analyzing sales data, track popularity, customer behavior, and other metrics, I aim to uncover patterns and make informed recommendations for business optimization.

## Query Writing Strategies
Throughout the project, I will adopt several strategies to write effective SQL queries:
- **Iterative Approach**: I will write queries in stages, testing and validating each step to ensure accuracy.
- **Debugging Techniques**: If faced with unexpected results, I will break down queries into smaller components for debugging.
- **Data Validation**: I will use separate queries to validate underlying data and calculations, ensuring the integrity of my analyses.

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

# Query the database to get a list of all tables in db

In [3]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

In [4]:
%%sql
SELECT *
FROM genre
Limit 10;

Done.


[(1, 'Rock'),
 (2, 'Jazz'),
 (3, 'Metal'),
 (4, 'Alternative & Punk'),
 (5, 'Rock And Roll'),
 (6, 'Blues'),
 (7, 'Latin'),
 (8, 'Reggae'),
 (9, 'Pop'),
 (10, 'Soundtrack')]

## Which genres sell best in the USA?
### Write a query that returns each genre, with the number of tracks sold in the USA:
- in absolute numbers
- in percentages.

In [5]:
%%sql
SELECT * 
FROM invoice
where billing_country = 'USA'
Limit 10;

Done.


[(1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83),
 (9, 18, '2017-01-18 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 8.91),
 (14, 25, '2017-01-23 00:00:00', '319 N. Frances Street', 'Madison', 'WI', 'USA', '53703', 0.99),
 (17, 20, '2017-01-28 00:00:00', '541 Del Medio Avenue', 'Mountain View', 'CA', 'USA', '94040-111', 10.89),
 (18, 24, '2017-02-02 00:00:00', '162 E Superior Street', 'Chicago', 'IL', 'USA', '60611', 4.95),
 (42, 19, '2017-03-18 00:00:00', '1 Infinite Loop', 'Cupertino', 'CA', 'USA', '95014', 8.91),
 (43, 22, '2017-03-19 00:00:00', '120 S Orange Ave', 'Orlando', 'FL', 'USA', '32801', 6.93),
 (44, 21, '2017-03-20 00:00:00', '801 W 4th Street', 'Reno', 'NV', 'USA', '89503', 11.879999999999999)]

In [12]:
%%sql
SELECT *
from invoice_line
limit 10

Done.


[(1, 1, 1158, 0.99, 1),
 (2, 1, 1159, 0.99, 1),
 (3, 1, 1160, 0.99, 1),
 (4, 1, 1161, 0.99, 1),
 (5, 1, 1162, 0.99, 1),
 (6, 1, 1163, 0.99, 1),
 (7, 1, 1164, 0.99, 1),
 (8, 1, 1165, 0.99, 1),
 (9, 1, 1166, 0.99, 1),
 (10, 1, 1167, 0.99, 1)]

In [14]:
%%sql
SELECT invoice_id, count(*)
FROM invoice_line
group by 1
order by count(*) desc
Limit 50;

Done.


[(183, 24),
 (31, 20),
 (92, 20),
 (362, 20),
 (526, 20),
 (276, 19),
 (54, 18),
 (99, 18),
 (141, 18),
 (175, 18),
 (223, 18),
 (582, 18),
 (604, 18),
 (5, 17),
 (56, 17),
 (71, 17),
 (112, 17),
 (114, 17),
 (273, 17),
 (282, 17),
 (320, 17),
 (344, 17),
 (389, 17),
 (398, 17),
 (449, 17),
 (478, 17),
 (479, 17),
 (1, 16),
 (237, 16),
 (239, 16),
 (336, 16),
 (468, 16),
 (506, 16),
 (528, 16),
 (73, 15),
 (142, 15),
 (305, 15),
 (334, 15),
 (335, 15),
 (413, 15),
 (427, 15),
 (454, 15),
 (457, 15),
 (69, 14),
 (79, 14),
 (153, 14),
 (281, 14),
 (372, 14),
 (404, 14),
 (488, 14)]

In [15]:
%%sql
WITH usa AS (
    SELECT 
        g.name as genre,
        ROUND(SUM(il.unit_price), 2) AS total,
        COUNT(distinct invoice_line_id) AS no_sold
    FROM 
        invoice AS i
        JOIN invoice_line AS il USING (invoice_id)
        JOIN playlist_track AS p ON il.track_id = p.track_id
        JOIN track AS t ON p.track_id = t.track_id
        JOIN genre AS g ON t.genre_id = g.genre_id
    WHERE 
        billing_country = 'USA'
    GROUP BY 
        g.name
    ORDER BY 
        total DESC
)

SELECT *
FROM usa;


Done.


[('Rock', 1371.15, 561),
 ('Alternative & Punk', 287.1, 130),
 ('Metal', 287.1, 124),
 ('R&B/Soul', 115.83, 53),
 ('Blues', 101.97, 36),
 ('Alternative', 71.28, 35),
 ('Hip Hop/Rap', 59.4, 20),
 ('Latin', 53.46, 22),
 ('Pop', 43.56, 22),
 ('Jazz', 28.71, 14),
 ('Easy Listening', 25.74, 13),
 ('Classical', 19.8, 4),
 ('Reggae', 17.82, 6),
 ('Electronica/Dance', 13.86, 5),
 ('Heavy Metal', 5.94, 3),
 ('Soundtrack', 4.95, 2),
 ('TV Shows', 1.98, 1)]

In [20]:
%%sql
WITH usa AS (
    SELECT 
       g.name as genre, 
       COUNT(DISTINCT invoice_line_id) as no_tracks_sold
    FROM 
        invoice AS i
        JOIN invoice_line AS il
        ON i.invoice_id = il.invoice_id
        JOIN customer c on i.customer_id = c.customer_id
        JOIN playlist_track AS p 
        ON il.track_id = p.track_id
        JOIN track AS t ON p.track_id = t.track_id
        JOIN genre AS g ON t.genre_id = g.genre_id
    WHERE 
        c.country = 'USA'
    GROUP BY 
        g.name
    ORDER BY 
        no_tracks_sold DESC
)

SELECT *,
    round(no_tracks_sold * 100.0/(SELECT SUM(no_tracks_sold) FROM usa),1) as perct
FROM usa;


Done.


[('Rock', 561, 53.4),
 ('Alternative & Punk', 130, 12.4),
 ('Metal', 124, 11.8),
 ('R&B/Soul', 53, 5.0),
 ('Blues', 36, 3.4),
 ('Alternative', 35, 3.3),
 ('Latin', 22, 2.1),
 ('Pop', 22, 2.1),
 ('Hip Hop/Rap', 20, 1.9),
 ('Jazz', 14, 1.3),
 ('Easy Listening', 13, 1.2),
 ('Reggae', 6, 0.6),
 ('Electronica/Dance', 5, 0.5),
 ('Classical', 4, 0.4),
 ('Heavy Metal', 3, 0.3),
 ('Soundtrack', 2, 0.2),
 ('TV Shows', 1, 0.1)]

### Extract the columns I need to find the employee with the highest sells within the company and any other relevant information

In [10]:
%%sql
with top_employees as
(SELECT e.employee_id, 
        e.first_name ||' ' || e.last_name as employee,
        e.title,
        e.reports_to,
        SUBSTR(e.hire_date,1,10) AS hire_date, e.hire_date string,
        e.country, i.total
FROM employee e
JOIN customer c
ON e.employee_id = c.support_rep_id
JOIN invoice i
ON c.customer_id = i.customer_id)

SELECT employee_id, employee, title, reports_to, hire_date, country,
    round(sum(total),2) as total_sales
FROM top_employees
GROUP BY 1,2,3,4,5,6;



Done.


[(3, 'Jane Peacock', 'Sales Support Agent', 2, '2017-04-01', 'Canada', 1731.51),
 (4, 'Margaret Park', 'Sales Support Agent', 2, '2017-05-03', 'Canada', 1584.0),
 (5, 'Steve Johnson', 'Sales Support Agent', 2, '2017-10-17', 'Canada', 1393.92)]


# Analyzing Sales by Country
* total number of customers
* total value of sales
* average value of sales per customer
* average order value

In [11]:
%%sql
select count(distinct customer_id) AS distinct_customers
from customer
limit 10

Done.


[(59,)]

In [12]:
%%sql
SELECT COUNT(DISTINCT c.country) no_countries
from customer c
LEFT JOIN invoice i
USING (customer_id)

Done.


[(24,)]

In [13]:
%%sql
SELECT COUNT(*) no_transactions
from customer c
LEFT JOIN invoice i
USING (customer_id)

Done.


[(614,)]

In [14]:
%%sql
select c.country, round(sum(total),2) as total_sales_per_country
from customer c
LEFT JOIN invoice i
USING (customer_id)
GROUP BY c.country
ORDER BY round(sum(total),2) DESC


Done.


[('USA', 1040.49),
 ('Canada', 535.59),
 ('Brazil', 427.68),
 ('France', 389.07),
 ('Germany', 334.62),
 ('Czech Republic', 273.24),
 ('United Kingdom', 245.52),
 ('Portugal', 185.13),
 ('India', 183.15),
 ('Ireland', 114.84),
 ('Spain', 98.01),
 ('Chile', 97.02),
 ('Australia', 81.18),
 ('Finland', 79.2),
 ('Hungary', 78.21),
 ('Poland', 76.23),
 ('Sweden', 75.24),
 ('Norway', 72.27),
 ('Austria', 69.3),
 ('Netherlands', 65.34),
 ('Belgium', 60.39),
 ('Italy', 50.49),
 ('Argentina', 39.6),
 ('Denmark', 37.62)]

In [15]:
%%sql
SELECT c.country, COUNT(*) as orders_per_country
from customer c
LEFT JOIN invoice i
USING (customer_id)
GROUP BY 1
ORDER BY 2 DESC

Done.


[('USA', 131),
 ('Canada', 76),
 ('Brazil', 61),
 ('France', 50),
 ('Germany', 41),
 ('Czech Republic', 30),
 ('Portugal', 29),
 ('United Kingdom', 28),
 ('India', 21),
 ('Chile', 13),
 ('Ireland', 13),
 ('Finland', 11),
 ('Spain', 11),
 ('Australia', 10),
 ('Denmark', 10),
 ('Hungary', 10),
 ('Netherlands', 10),
 ('Poland', 10),
 ('Sweden', 10),
 ('Austria', 9),
 ('Italy', 9),
 ('Norway', 9),
 ('Belgium', 7),
 ('Argentina', 5)]

## How many customers are in each country?
* Where a country has only one customer, collect them into an "Other" group.

In [16]:
%%sql
SELECT c.country, count(distinct c.customer_id) as no_of_customers
from customer c
LEFT JOIN invoice i
USING (customer_id)
group by 1
order by 2 desc

Done.


[('USA', 13),
 ('Canada', 8),
 ('Brazil', 5),
 ('France', 5),
 ('Germany', 4),
 ('United Kingdom', 3),
 ('Czech Republic', 2),
 ('India', 2),
 ('Portugal', 2),
 ('Argentina', 1),
 ('Australia', 1),
 ('Austria', 1),
 ('Belgium', 1),
 ('Chile', 1),
 ('Denmark', 1),
 ('Finland', 1),
 ('Hungary', 1),
 ('Ireland', 1),
 ('Italy', 1),
 ('Netherlands', 1),
 ('Norway', 1),
 ('Poland', 1),
 ('Spain', 1),
 ('Sweden', 1)]

### Now group the country column

In [17]:
%%sql
WITH country_customer as(
    SELECT c.country, 
           count(distinct c.customer_id) as no_customers
from customer c
LEFT JOIN invoice i
USING (customer_id)
group by 1
order by 2 desc)

select country, sum(no_customers) AS no_of_customers
from (select case when no_customers <= 1 then 'Other' else 
    country end as country, no_customers
from country_customer) as collection
group by 1
ORDER BY sum(no_customers) DESC
    

Done.


[('Other', 15),
 ('USA', 13),
 ('Canada', 8),
 ('Brazil', 5),
 ('France', 5),
 ('Germany', 4),
 ('United Kingdom', 3),
 ('Czech Republic', 2),
 ('India', 2),
 ('Portugal', 2)]

In [18]:
%%sql
SELECT
    c.country,
    count(c.customer_id) AS order_per_country, 
    ROUND(SUM(total)) AS total_spend,
    COUNT(DISTINCT c.customer_id) AS no_of_customers
FROM customer c
LEFT JOIN invoice i
USING (customer_id)
GROUP BY 1

Done.


[('Argentina', 5, 40.0, 1),
 ('Australia', 10, 81.0, 1),
 ('Austria', 9, 69.0, 1),
 ('Belgium', 7, 60.0, 1),
 ('Brazil', 61, 428.0, 5),
 ('Canada', 76, 536.0, 8),
 ('Chile', 13, 97.0, 1),
 ('Czech Republic', 30, 273.0, 2),
 ('Denmark', 10, 38.0, 1),
 ('Finland', 11, 79.0, 1),
 ('France', 50, 389.0, 5),
 ('Germany', 41, 335.0, 4),
 ('Hungary', 10, 78.0, 1),
 ('India', 21, 183.0, 2),
 ('Ireland', 13, 115.0, 1),
 ('Italy', 9, 50.0, 1),
 ('Netherlands', 10, 65.0, 1),
 ('Norway', 9, 72.0, 1),
 ('Poland', 10, 76.0, 1),
 ('Portugal', 29, 185.0, 2),
 ('Spain', 11, 98.0, 1),
 ('Sweden', 10, 75.0, 1),
 ('USA', 131, 1040.0, 13),
 ('United Kingdom', 28, 246.0, 3)]

In [19]:
%%sql
WITH condensed AS(SELECT
    c.country,
    count(c.customer_id) AS order_per_country, 
    ROUND(SUM(total)) AS total_spend,
    COUNT(DISTINCT c.customer_id) AS no_of_customers
FROM customer c
LEFT JOIN invoice i
USING (customer_id)
GROUP BY 1)

SELECT
    CASE WHEN no_of_customers <= 1 then 'Other' ELSE country END AS country,
    SUM(order_per_country) AS orders_per_country,
    SUM(total_spend) AS total_spend,
    SUM(no_of_customers) AS no_of_customers
FROM condensed
GROUP BY 1

Done.


[('Brazil', 61, 428.0, 5),
 ('Canada', 76, 536.0, 8),
 ('Czech Republic', 30, 273.0, 2),
 ('France', 50, 389.0, 5),
 ('Germany', 41, 335.0, 4),
 ('India', 21, 183.0, 2),
 ('Other', 147, 1093.0, 15),
 ('Portugal', 29, 185.0, 2),
 ('USA', 131, 1040.0, 13),
 ('United Kingdom', 28, 246.0, 3)]

### Force 'Other' to the bottom of your results.

In [20]:
%%sql
WITH condensed AS(SELECT
    c.country,
    count(c.customer_id) AS order_per_country, 
    ROUND(SUM(total),2) AS total_spend,
    COUNT(DISTINCT c.customer_id) AS no_of_customers
FROM customer c
LEFT JOIN invoice i
USING (customer_id)
GROUP BY 1),

final_countries AS(SELECT
    CASE WHEN no_of_customers <= 1 then 'Other' ELSE country END AS country,
    SUM(no_of_customers) AS no_of_customers,
    SUM(order_per_country) AS orders_per_country,
    SUM(total_spend) AS total_spend,
    CASE WHEN no_of_customers <= 1 THEN 1 ELSE 0 END AS sort
FROM condensed
GROUP BY 1)

SELECT country, no_of_customers, orders_per_country, total_spend,
    ROUND(total_spend/no_of_customers,2) AS avg_value_per_customer,
    ROUND(total_spend/orders_per_country,2) AS average_order_value
FROM final_countries
ORDER BY sort ASC, total_spend DESC


Done.


[('USA', 13, 131, 1040.49, 80.04, 7.94),
 ('Canada', 8, 76, 535.59, 66.95, 7.05),
 ('Brazil', 5, 61, 427.68, 85.54, 7.01),
 ('France', 5, 50, 389.07, 77.81, 7.78),
 ('Germany', 4, 41, 334.62, 83.66, 8.16),
 ('Czech Republic', 2, 30, 273.24, 136.62, 9.11),
 ('United Kingdom', 3, 28, 245.52, 81.84, 8.77),
 ('Portugal', 2, 29, 185.13, 92.56, 6.38),
 ('India', 2, 21, 183.15, 91.58, 8.72),
 ('Other', 15, 147, 1094.94, 73.0, 7.45)]

# Albums vs Individual Tracks
* We have been asked to find out what percentage of purchases are individual tracks vs whole albums, so that management can use this data to understand the effect this decision might have on overall revenue.

In [33]:
%%sql
select count(distinct album_id)
from album

Done.


[(347,)]

### How many tracks are in each album?
* there are 347 distinct albums in the database
* 3503 tracks

In [47]:
%%sql
SELECT
    al.album_id,
    count(track_id)no_tracks_per_album,
    min(tr.track_id)first_track_in_album
FROM album al
LEFT JOIN track tr
ON al.album_id = tr.album_id
GROUP BY 1

Done.


[(1, 10, 1),
 (2, 1, 2),
 (3, 3, 3),
 (4, 8, 15),
 (5, 15, 23),
 (6, 13, 38),
 (7, 12, 51),
 (8, 14, 63),
 (9, 8, 77),
 (10, 14, 85),
 (11, 12, 99),
 (12, 12, 111),
 (13, 8, 123),
 (14, 13, 131),
 (15, 5, 144),
 (16, 7, 149),
 (17, 10, 156),
 (18, 17, 166),
 (19, 11, 183),
 (20, 11, 194),
 (21, 18, 205),
 (22, 3, 223),
 (23, 34, 226),
 (24, 23, 246),
 (25, 13, 269),
 (26, 17, 282),
 (27, 14, 299),
 (28, 10, 313),
 (29, 14, 323),
 (30, 14, 337),
 (31, 9, 351),
 (32, 14, 360),
 (33, 17, 374),
 (34, 17, 391),
 (35, 11, 408),
 (36, 17, 419),
 (37, 20, 436),
 (38, 12, 456),
 (39, 21, 468),
 (40, 12, 489),
 (41, 14, 501),
 (42, 14, 529),
 (43, 7, 543),
 (44, 6, 550),
 (45, 14, 556),
 (46, 13, 570),
 (47, 14, 583),
 (48, 13, 597),
 (49, 10, 610),
 (50, 4, 620),
 (51, 22, 624),
 (52, 15, 646),
 (53, 14, 661),
 (54, 20, 675),
 (55, 20, 695),
 (56, 15, 715),
 (57, 15, 730),
 (58, 9, 745),
 (59, 7, 754),
 (60, 7, 761),
 (61, 11, 768),
 (62, 7, 779),
 (63, 12, 786),
 (64, 9, 798),
 (65, 9, 807),
 (66, 10, 816),
 (67, 16, 826),
 (68, 9, 842),
 (69, 13, 851),
 (70, 13, 864),
 (71, 14, 877),
 (72, 18, 891),
 (73, 30, 909),
 (74, 12, 923),
 (75, 14, 935),
 (76, 15, 949),
 (77, 11, 964),
 (78, 14, 975),
 (79, 10, 989),
 (80, 10, 999),
 (81, 11, 1009),
 (82, 13, 1020),
 (83, 24, 1033),
 (84, 16, 1057),
 (85, 14, 1073),
 (86, 15, 1087),
 (87, 3, 1102),
 (88, 12, 1121),
 (89, 13, 1133),
 (90, 12, 1146),
 (91, 16, 1158),
 (92, 14, 1174),
 (93, 13, 1188),
 (94, 11, 1201),
 (95, 12, 1212),
 (96, 11, 1224),
 (97, 10, 1235),
 (98, 11, 1245),
 (99, 12, 1256),
 (100, 9, 1268),
 (101, 10, 1277),
 (102, 18, 1287),
 (103, 10, 1305),
 (104, 10, 1315),
 (105, 10, 1325),
 (106, 9, 1335),
 (107, 8, 1344),
 (108, 10, 1352),
 (109, 9, 1362),
 (110, 8, 1371),
 (111, 8, 1379),
 (112, 8, 1387),
 (113, 11, 1395),
 (114, 8, 1406),
 (115, 20, 1414),
 (116, 10, 1434),
 (117, 11, 1444),
 (118, 11, 1455),
 (119, 13, 1466),
 (120, 17, 1479),
 (121, 10, 1496),
 (122, 14, 1506),
 (123, 12, 1520),
 (124, 14, 1532),
 (125, 16, 1546),
 (126, 15, 1562),
 (127, 10, 1577),
 (128, 8, 1587),
 (129, 8, 1595),
 (130, 7, 1603),
 (131, 8, 1610),
 (132, 9, 1618),
 (133, 9, 1627),
 (134, 10, 1636),
 (135, 9, 1646),
 (136, 7, 1655),
 (137, 5, 1662),
 (138, 4, 1667),
 (139, 15, 1671),
 (140, 16, 1686),
 (141, 57, 1702),
 (142, 14, 1717),
 (143, 14, 1718),
 (144, 10, 1745),
 (145, 18, 1755),
 (146, 18, 1773),
 (147, 10, 1791),
 (148, 12, 1801),
 (149, 16, 1813),
 (150, 10, 1829),
 (151, 14, 1839),
 (152, 8, 1853),
 (153, 13, 1861),
 (154, 8, 1874),
 (155, 11, 1882),
 (156, 9, 1893),
 (157, 14, 1902),
 (158, 13, 1916),
 (159, 13, 1929),
 (160, 15, 1942),
 (161, 12, 1957),
 (162, 17, 1969),
 (163, 17, 1986),
 (164, 12, 2003),
 (165, 15, 2015),
 (166, 14, 2030),
 (167, 21, 2044),
 (168, 12, 2065),
 (169, 16, 2077),
 (170, 1, 2093),
 (171, 2, 2094),
 (172, 1, 2096),
 (173, 2, 2097),
 (174, 14, 2099),
 (175, 12, 2113),
 (176, 14, 2125),
 (177, 10, 2139),
 (178, 16, 2149),
 (179, 13, 2165),
 (180, 15, 2178),
 (181, 11, 2193),
 (182, 12, 2204),
 (183, 9, 2229),
 (184, 16, 2238),
 (185, 17, 2254),
 (186, 11, 2271),
 (187, 11, 2282),
 (188, 11, 2285),
 (189, 14, 2296),
 (190, 16, 2318),
 (191, 10, 2334),
 (192, 14, 2344),
 (193, 17, 2358),
 (194, 16, 2375),
 (195, 15, 2391),
 (196, 14, 2406),
 (197, 8, 2420),
 (198, 6, 2428),
 (199, 12, 2449),
 (200, 11, 2461),
 (201, 16, 2472),
 (202, 18, 2488),
 (203, 17, 2506),
 (204, 9, 2523),
 (205, 10, 2532),
 (206, 12, 2542),
 (207, 11, 2554),
 (208, 7, 2565),
 (209, 10, 2572),
 (210, 9, 2582),
 (211, 18, 2591),
 (212, 12, 2609),
 (213, 18, 2621),
 (214, 11, 2639),
 (215, 14, 2650),
 (216, 12, 2664),
 (217, 14, 2676),
 (218, 15, 2690),
 (219, 15, 2705),
 (220, 11, 2720),
 (221, 20, 2731),
 (222, 15, 2751),
 (223, 15, 2766),
 (224, 22, 2781),
 (225, 16, 2803),
 (226, 1, 2819),
 (227, 19, 2820),
 (228, 23, 2839),
 (229, 26, 2857),
 (230, 25, 2858),
 (231, 24, 2859),
 (232, 12, 2926),
 (233, 11, 2938),
 (234, 15, 2949),
 (235, 11, 2964),
 (236, 12, 2975),
 (237,

In [31]:
%%sql
select
    i.invoice_id,
    i.customer_id,
    i.invoice_date,
    il.track_id,
    tr.name,
    al.album_id,
    al.title
from invoice i
left join invoice_line il
on i.invoice_id = il.invoice_id
left join track tr
on tr.track_id = il.track_id
left join album al
on al.album_id = tr.album_id
limit 10

Done.


[(1, 18, '2017-01-03 00:00:00', 1158, 'Right Next Door to Hell', 91, 'Use Your Illusion I'),
 (1, 18, '2017-01-03 00:00:00', 1159, "Dust N' Bones", 91, 'Use Your Illusion I'),
 (1, 18, '2017-01-03 00:00:00', 1160, 'Live and Let Die', 91, 'Use Your Illusion I'),
 (1, 18, '2017-01-03 00:00:00', 1161, "Don't Cry (Original)", 91, 'Use Your Illusion I'),
 (1, 18, '2017-01-03 00:00:00', 1162, 'Perfect Crime', 91, 'Use Your Illusion I'),
 (1, 18, '2017-01-03 00:00:00', 1163, "You Ain't the First", 91, 'Use Your Illusion I'),
 (1, 18, '2017-01-03 00:00:00', 1164, 'Bad Obsession', 91, 'Use Your Illusion I'),
 (1, 18, '2017-01-03 00:00:00', 1165, 'Back off Bitch', 91, 'Use Your Illusion I'),
 (1, 18, '2017-01-03 00:00:00', 1166, "Double Talkin' Jive", 91, 'Use Your Illusion I'),
 (1, 18, '2017-01-03 00:00:00', 1167, 'November Rain', 91, 'Use Your Illusion I')]

In [37]:
%%sql
with first_tracks_bought as(SELECT
 invoice_id, count(*) no_of_tracks_bought,
    MIN(il.track_id) AS first_track_id
  FROM
    Invoice_Line il
  GROUP BY
    1),
album_tracks as (select al.album_id, count(track_id) tracks_per_album, min(tr.track_id) as first_track
from album al
left join track tr
on al.album_id = tr.album_id
group by 1)


select h.*,case when album_id is null and tracks_per_album is null and first_track is null then 'no' else 'yes' end album_bought
from(select *
from first_tracks_bought fb
left join 
album_tracks at
on fb.first_track_id = at.first_track
and fb.no_of_tracks_bought = at.tracks_per_album) h

Done.


[(1, 16, 1158, 91, 16, 1158, 'yes'),
 (2, 10, 201, None, None, None, 'no'),
 (3, 2, 2516, None, None, None, 'no'),
 (4, 8, 748, None, None, None, 'no'),
 (5, 17, 1986, 163, 17, 1986, 'yes'),
 (6, 2, 30, None, None, None, 'no'),
 (7, 11, 42, None, None, None, 'no'),
 (8, 10, 81, None, None, None, 'no'),
 (9, 9, 196, None, None, None, 'no'),
 (10, 2, 2663, None, None, None, 'no'),
 (11, 11, 610, None, None, None, 'no'),
 (12, 4, 92, None, None, None, 'no'),
 (13, 1, 2553, None, None, None, 'no'),
 (14, 1, 541, None, None, None, 'no'),
 (15, 4, 807, None, None, None, 'no'),
 (16, 9, 16, None, None, None, 'no'),
 (17, 11, 55, None, None, None, 'no'),
 (18, 5, 1027, None, None, None, 'no'),
 (19, 7, 105, None, None, None, 'no'),
 (20, 10, 60, None, None, None, 'no'),
 (21, 9, 13, None, None, None, 'no'),
 (22, 2, 383, None, None, None, 'no'),
 (23, 10, 1, 1, 10, 1, 'yes'),
 (24, 12, 1146, 90, 12, 1146, 'yes'),
 (25, 8, 32, None, None, None, 'no'),
 (26, 2, 85, None, None, None, 'no'),
 (27, 8, 1280, None, None, None, 'no'),
 (28, 3, 36, None, None, None, 'no'),
 (29, 6, 479, None, None, None, 'no'),
 (30, 11, 33, None, None, None, 'no'),
 (31, 20, 2731, 221, 20, 2731, 'yes'),
 (32, 13, 38, 6, 13, 38, 'yes'),
 (33, 4, 2542, None, None, None, 'no'),
 (34, 1, 610, None, None, None, 'no'),
 (35, 5, 6, None, None, None, 'no'),
 (36, 7, 153, None, None, None, 'no'),
 (37, 10, 698, None, None, None, 'no'),
 (38, 7, 154, None, None, None, 'no'),
 (39, 8, 184, None, None, None, 'no'),
 (40, 10, 108, None, None, None, 'no'),
 (41, 8, 153, None, None, None, 'no'),
 (42, 9, 13, None, None, None, 'no'),
 (43, 7, 918, None, None, None, 'no'),
 (44, 12, 3052, 242, 12, 3052, 'yes'),
 (45, 8, 102, None, None, None, 'no'),
 (46, 11, 2639, 214, 11, 2639, 'yes'),
 (47, 6, 8, None, None, None, 'no'),
 (48, 4, 58, None, None, None, 'no'),
 (49, 5, 146, None, None, None, 'no'),
 (50, 7, 1511, None, None, None, 'no'),
 (51, 8, 34, None, None, None, 'no'),
 (52, 12, 2664, 216, 12, 2664, 'yes'),
 (53, 3, 207, None, None, None, 'no'),
 (54, 18, 1773, 146, 18, 1773, 'yes'),
 (55, 10, 155, None, None, None, 'no'),
 (56, 17, 2254, 185, 17, 2254, 'yes'),
 (57, 1, 2948, None, None, None, 'no'),
 (58, 6, 1029, None, None, None, 'no'),
 (59, 11, 198, None, None, None, 'no'),
 (60, 8, 46, None, None, None, 'no'),
 (61, 12, 2609, 212, 12, 2609, 'yes'),
 (62, 3, 97, None, None, None, 'no'),
 (63, 6, 6, None, None, None, 'no'),
 (64, 13, 38, 6, 13, 38, 'yes'),
 (65, 2, 539, None, None, None, 'no'),
 (66, 7, 442, None, None, None, 'no'),
 (67, 11, 9, None, None, None, 'no'),
 (68, 11, 61, None, None, None, 'no'),
 (69, 14, 85, 10, 14, 85, 'yes'),
 (70, 1, 2307, None, None, None, 'no'),
 (71, 17, 2254, 185, 17, 2254, 'yes'),
 (72, 3, 30, None, None, None, 'no'),
 (73, 15, 2391, 195, 15, 2391, 'yes'),
 (74, 2, 1743, None, None, None, 'no'),
 (75, 4, 22, None, None, None, 'no'),
 (76, 13, 1020, 82, 13, 1020, 'yes'),
 (77, 6, 15, None, None, None, 'no'),
 (78, 9, 33, None, None, None, 'no'),
 (79, 14, 1902, 157, 14, 1902, 'yes'),
 (80, 1, 3400, None, None, None, 'no'),
 (81, 6, 545, None, None, None, 'no'),
 (82, 7, 54, None, None, None, 'no'),
 (83, 4, 1163, None, None, None, 'no'),
 (84, 12, 3455, 321, 12, 3455, 'yes'),
 (85, 5, 1298, None, None, None, 'no'),
 (86, 1, 2291, None, None, None, 'no'),
 (87, 8, 1129, None, None, None, 'no'),
 (88, 5, 1995, None, None, None, 'no'),
 (89, 2, 1784, None, None, None, 'no'),
 (90, 11, 2554, 207, 11, 2554, 'yes'),
 (91, 11, 94, None, None, None, 'no'),
 (92, 20, 1414, 115, 20, 1414, 'yes'),
 (93, 10, 20, None, None, None, 'no'),
 (94, 8, 50, None, None, None, 'no'),
 (95, 11, 27, None, None, None, 'no'),
 (96, 6, 453, None, None, None, 'no'),
 (97, 6, 16, None, None, None, 'no'),
 (98, 10, 419, None, None, None, 'no'),
 (99, 18, 2591, 211, 18, 2591, 'yes'),
 (100, 6, 996, None, None, None, 'no'),
 (101, 4, 83, None, None, None, 'no'),
 (102, 10, 33, None, None, None, 'no'),
 (103, 6, 1029, None, None, None, 'no'),
 (104, 2, 1508, N

In [51]:
%%sql
select al.album_id,al.title,tr.track_id,tr.name 
from album al
left join track tr
on al.album_id = tr.album_id
limit 20

Done.


[(1, 'For Those About To Rock We Salute You', 1, 'For Those About To Rock (We Salute You)'),
 (1, 'For Those About To Rock We Salute You', 6, 'Put The Finger On You'),
 (1, 'For Those About To Rock We Salute You', 7, "Let's Get It Up"),
 (1, 'For Those About To Rock We Salute You', 8, 'Inject The Venom'),
 (1, 'For Those About To Rock We Salute You', 9, 'Snowballed'),
 (1, 'For Those About To Rock We Salute You', 10, 'Evil Walks'),
 (1, 'For Those About To Rock We Salute You', 11, 'C.O.D.'),
 (1, 'For Those About To Rock We Salute You', 12, 'Breaking The Rules'),
 (1, 'For Those About To Rock We Salute You', 13, 'Night Of The Long Knives'),
 (1, 'For Those About To Rock We Salute You', 14, 'Spellbound'),
 (2, 'Balls to the Wall', 2, 'Balls to the Wall'),
 (3, 'Restless and Wild', 3, 'Fast As a Shark'),
 (3, 'Restless and Wild', 4, 'Restless and Wild'),
 (3, 'Restless and Wild', 5, 'Princess of the Dawn'),
 (4, 'Let There Be Rock', 15, 'Go Down'),
 (4, 'Let There Be Rock', 16, 'Dog Eat Dog'),
 (4, 'Let There Be Rock', 17, 'Let There Be Rock'),
 (4, 'Let There Be Rock', 18, 'Bad Boy Boogie'),
 (4, 'Let There Be Rock', 19, 'Problem Child'),
 (4, 'Let There Be Rock', 20, 'Overdose')]

In [45]:
%%sql
with first_tracks_bought as(SELECT
 invoice_id, count(*) no_of_tracks_bought,
    MIN(il.track_id) AS first_track_id
  FROM
    Invoice_Line il
  GROUP BY
    1),
album_tracks as (select al.album_id, count(track_id) tracks_per_album, min(tr.track_id) as first_track
from album al
left join track tr
on al.album_id = tr.album_id
group by 1)

select 
    album_bought,
    count(*) no_invoices,
    round(cast(count(*) as float)/(select count(*)from first_tracks_bought),2) Percentage_of_invoices
from(select h.*,case when tracks_per_album is null and first_track is null then 'no' else 'yes' end album_bought
from(select *
from first_tracks_bought fb
left join 
album_tracks at
on fb.first_track_id = at.first_track
and fb.no_of_tracks_bought = at.tracks_per_album) h)j
group by 1

Done.


[('no', 498, 0.81), ('yes', 116, 0.19)]

* first_tracks_bought: This CTE calculates the first track bought in each invoice and counts the total number of tracks bought in each invoice.
* album_tracks: This CTE calculates the number of tracks per album and identifies the first track for each album.
* The main query joins the results of the two CTEs, first_tracks_bought and album_tracks, based on the condition that the first track bought in each invoice matches the first track of an album, and the number of tracks bought equals the total tracks in the album.
* It then uses a CASE statement to determine whether an album was bought. If the join conditions fail, indicating that either the album ID or the tracks per album or the first track is null, it implies that the album was not bought, and it assigns 'no' to the album_bought column; otherwise, it assigns 'yes'.
* The result set contains all the columns from the joined CTEs along with the album_bought column indicating whether an album was bought or not.

# Analysis of Album vs. Individual Track Purchases

## Findings:
Based on the results of the SQL query:

- Out of the total number of invoices analyzed, 498 invoices (approximately 81%) were categorized as not purchasing albums, while 116 invoices (approximately 19%) were classified as purchasing albums.
- This indicates that a significant majority of the invoices in the Chinook store are comprised of individual track purchases rather than full album purchases.

## Recommendations:
Based on these findings:

- **Optimize Individual Track Sales**: Given that a majority of customers are opting for individual track purchases, the store may focus on optimizing revenue from individual track sales.
- **Promote Popular Tracks**: Focus on promoting popular individual tracks through targeted marketing campaigns, curated playlists, and recommendations.
- **Explore Artist Partnerships**: Explore partnerships with artists to promote singles and popular tracks, potentially leading to increased revenue.
- **Customer Preferences Analysis**: Conduct further analysis to understand customer preferences, listening habits, and trends to tailor the music catalog and promotions accordingly.